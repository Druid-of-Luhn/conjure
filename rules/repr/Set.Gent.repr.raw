RuleRepr
  { reprFilename = "rules/repr/Set.Gent.repr"
  , reprName = "Gent"
  , reprTemplate = DomainMatrix
                     { index = Identifier "__rules_repr_Set_Gent_repr_t"
                     , element = DomainIntegerFromTo (
                                   Just ( ValueInteger 0 ) ) (
                                   Just ( Identifier "__rules_repr_Set_Gent_repr_mx" ) )
                     }
  , reprPrologueStructural = Just
                               ExprQuantifier
                                 { quanName = Identifier "forall"
                                 , quanVar = Identifier "__rules_repr_Set_Gent_repr_i"
                                 , quanOver = Identifier "__rules_repr_Set_Gent_repr_t"
                                 , quanGuard = Nothing
                                 , quanBody = GenericNode
                                                Or
                                                [ GenericNode
                                                    Eq
                                                    [ GenericNode
                                                        Index
                                                        [ Identifier "refn"
                                                        , Identifier "__rules_repr_Set_Gent_repr_i"
                                                        ]
                                                    , ValueInteger 0
                                                    ]
                                                , GenericNode
                                                    Eq
                                                    [ GenericNode
                                                        Index
                                                        [ Identifier "refn"
                                                        , Identifier "__rules_repr_Set_Gent_repr_i"
                                                        ]
                                                    , GenericNode
                                                        Plus
                                                        [ ValueInteger 1
                                                        , ExprQuantifier
                                                            { quanName = Identifier "sum"
                                                            , quanVar = Identifier "__rules_repr_Set_Gent_repr_j"
                                                            , quanOver = Identifier "__rules_repr_Set_Gent_repr_t"
                                                            , quanGuard = Just (
                                                                            GenericNode
                                                                              And
                                                                              [ GenericNode
                                                                                  Lt
                                                                                  [ Identifier "__rules_repr_Set_Gent_repr_j"
                                                                                  , Identifier "__rules_repr_Set_Gent_repr_i"
                                                                                  ]
                                                                              , GenericNode
                                                                                  Gt
                                                                                  [ GenericNode
                                                                                      Index
                                                                                      [ Identifier "refn"
                                                                                      , Identifier "__rules_repr_Set_Gent_repr_j"
                                                                                      ]
                                                                                  , ValueInteger 0
                                                                                  ]
                                                                              ] )
                                                            , quanBody = ValueInteger 1
                                                            }
                                                        ]
                                                    ]
                                                ]
                                 }
  , reprPrologueWheres = [ GenericNode
                             HasType
                             [ Identifier "__rules_repr_Set_Gent_repr_t"
                             , DomainIntegerList []
                             ]
                         ]
  , reprPrologueBindings = []
  , reprCases = [ RuleReprCase
                    { reprCasePattern = DomainSet
                                          { size = Nothing
                                          , minSize = Nothing
                                          , maxSize = Nothing
                                          , attrDontCare = False
                                          , element = Identifier "__rules_repr_Set_Gent_repr_t"
                                          , representation = Nothing
                                          }
                    , reprCaseStructural = Nothing
                    , reprCaseWheres = []
                    , reprCaseBindings = [ ( Letting
                                           , "__rules_repr_Set_Gent_repr_mx"
                                           , GenericNode
                                               Image
                                               [ Identifier "domSize"
                                               , Identifier "__rules_repr_Set_Gent_repr_t"
                                               ]
                                           )
                                         ]
                    }
                , RuleReprCase
                    { reprCasePattern = DomainSet
                                          { size = Just ( Identifier "__rules_repr_Set_Gent_repr_n" )
                                          , minSize = Nothing
                                          , maxSize = Nothing
                                          , attrDontCare = False
                                          , element = Identifier "__rules_repr_Set_Gent_repr_t"
                                          , representation = Nothing
                                          }
                    , reprCaseStructural = Just (
                                             GenericNode
                                               Eq
                                               [ ExprQuantifier
                                                   { quanName = Identifier "sum"
                                                   , quanVar = Identifier "__rules_repr_Set_Gent_repr_i"
                                                   , quanOver = Identifier "__rules_repr_Set_Gent_repr_t"
                                                   , quanGuard = Just (
                                                                   GenericNode
                                                                     Gt
                                                                     [ GenericNode
                                                                         Index
                                                                         [ Identifier "refn"
                                                                         , Identifier "__rules_repr_Set_Gent_repr_i"
                                                                         ]
                                                                     , ValueInteger 0
                                                                     ] )
                                                   , quanBody = ValueInteger 1
                                                   }
                                               , Identifier "__rules_repr_Set_Gent_repr_n"
                                               ] )
                    , reprCaseWheres = []
                    , reprCaseBindings = [ ( Letting
                                           , "__rules_repr_Set_Gent_repr_mx"
                                           , GenericNode
                                               Image
                                               [ Identifier "domSize"
                                               , Identifier "__rules_repr_Set_Gent_repr_t"
                                               ]
                                           )
                                         ]
                    }
                , RuleReprCase
                    { reprCasePattern = DomainSet
                                          { size = Nothing
                                          , minSize = Just ( Identifier "__rules_repr_Set_Gent_repr_mn" )
                                          , maxSize = Nothing
                                          , attrDontCare = False
                                          , element = Identifier "__rules_repr_Set_Gent_repr_t"
                                          , representation = Nothing
                                          }
                    , reprCaseStructural = Just (
                                             GenericNode
                                               Geq
                                               [ ExprQuantifier
                                                   { quanName = Identifier "sum"
                                                   , quanVar = Identifier "__rules_repr_Set_Gent_repr_i"
                                                   , quanOver = Identifier "__rules_repr_Set_Gent_repr_t"
                                                   , quanGuard = Just (
                                                                   GenericNode
                                                                     Gt
                                                                     [ GenericNode
                                                                         Index
                                                                         [ Identifier "refn"
                                                                         , Identifier "__rules_repr_Set_Gent_repr_i"
                                                                         ]
                                                                     , ValueInteger 0
                                                                     ] )
                                                   , quanBody = ValueInteger 1
                                                   }
                                               , Identifier "__rules_repr_Set_Gent_repr_mn"
                                               ] )
                    , reprCaseWheres = []
                    , reprCaseBindings = [ ( Letting
                                           , "__rules_repr_Set_Gent_repr_mx"
                                           , GenericNode
                                               Image
                                               [ Identifier "domSize"
                                               , Identifier "__rules_repr_Set_Gent_repr_t"
                                               ]
                                           )
                                         ]
                    }
                , RuleReprCase
                    { reprCasePattern = DomainSet
                                          { size = Nothing
                                          , minSize = Nothing
                                          , maxSize = Just ( Identifier "__rules_repr_Set_Gent_repr_mx" )
                                          , attrDontCare = False
                                          , element = Identifier "__rules_repr_Set_Gent_repr_t"
                                          , representation = Nothing
                                          }
                    , reprCaseStructural = Just (
                                             GenericNode
                                               Leq
                                               [ ExprQuantifier
                                                   { quanName = Identifier "sum"
                                                   , quanVar = Identifier "__rules_repr_Set_Gent_repr_i"
                                                   , quanOver = Identifier "__rules_repr_Set_Gent_repr_t"
                                                   , quanGuard = Just (
                                                                   GenericNode
                                                                     Gt
                                                                     [ GenericNode
                                                                         Index
                                                                         [ Identifier "refn"
                                                                         , Identifier "__rules_repr_Set_Gent_repr_i"
                                                                         ]
                                                                     , ValueInteger 0
                                                                     ] )
                                                   , quanBody = ValueInteger 1
                                                   }
                                               , Identifier "__rules_repr_Set_Gent_repr_mx"
                                               ] )
                    , reprCaseWheres = []
                    , reprCaseBindings = []
                    }
                , RuleReprCase
                    { reprCasePattern = DomainSet
                                          { size = Just ( Identifier "__rules_repr_Set_Gent_repr_n" )
                                          , minSize = Just ( Identifier "__rules_repr_Set_Gent_repr_mn" )
                                          , maxSize = Nothing
                                          , attrDontCare = False
                                          , element = Identifier "__rules_repr_Set_Gent_repr_t"
                                          , representation = Nothing
                                          }
                    , reprCaseStructural = Just (
                                             GenericNode
                                               Eq
                                               [ ExprQuantifier
                                                   { quanName = Identifier "sum"
                                                   , quanVar = Identifier "__rules_repr_Set_Gent_repr_i"
                                                   , quanOver = Identifier "__rules_repr_Set_Gent_repr_t"
                                                   , quanGuard = Just (
                                                                   GenericNode
                                                                     Gt
                                                                     [ GenericNode
                                                                         Index
                                                                         [ Identifier "refn"
                                                                         , Identifier "__rules_repr_Set_Gent_repr_i"
                                                                         ]
                                                                     , ValueInteger 0
                                                                     ] )
                                                   , quanBody = ValueInteger 1
                                                   }
                                               , Identifier "__rules_repr_Set_Gent_repr_n"
                                               ] )
                    , reprCaseWheres = []
                    , reprCaseBindings = [ ( Letting
                                           , "__rules_repr_Set_Gent_repr_mx"
                                           , GenericNode
                                               Image
                                               [ Identifier "domSize"
                                               , Identifier "__rules_repr_Set_Gent_repr_t"
                                               ]
                                           )
                                         ]
                    }
                , RuleReprCase
                    { reprCasePattern = DomainSet
                                          { size = Just ( Identifier "__rules_repr_Set_Gent_repr_n" )
                                          , minSize = Nothing
                                          , maxSize = Just ( Identifier "__rules_repr_Set_Gent_repr_mx" )
                                          , attrDontCare = False
                                          , element = Identifier "__rules_repr_Set_Gent_repr_t"
                                          , representation = Nothing
                                          }
                    , reprCaseStructural = Just (
                                             GenericNode
                                               Eq
                                               [ ExprQuantifier
                                                   { quanName = Identifier "sum"
                                                   , quanVar = Identifier "__rules_repr_Set_Gent_repr_i"
                                                   , quanOver = Identifier "__rules_repr_Set_Gent_repr_t"
                                                   , quanGuard = Just (
                                                                   GenericNode
                                                                     Gt
                                                                     [ GenericNode
                                                                         Index
                                                                         [ Identifier "refn"
                                                                         , Identifier "__rules_repr_Set_Gent_repr_i"
                                                                         ]
                                                                     , ValueInteger 0
                                                                     ] )
                                                   , quanBody = ValueInteger 1
                                                   }
                                               , Identifier "__rules_repr_Set_Gent_repr_n"
                                               ] )
                    , reprCaseWheres = []
                    , reprCaseBindings = []
                    }
                , RuleReprCase
                    { reprCasePattern = DomainSet
                                          { size = Nothing
                                          , minSize = Just ( Identifier "__rules_repr_Set_Gent_repr_mn" )
                                          , maxSize = Just ( Identifier "__rules_repr_Set_Gent_repr_mx" )
                                          , attrDontCare = False
                                          , element = Identifier "__rules_repr_Set_Gent_repr_t"
                                          , representation = Nothing
                                          }
                    , reprCaseStructural = Just (
                                             GenericNode
                                               And
                                               [ GenericNode
                                                   Geq
                                                   [ ExprQuantifier
                                                       { quanName = Identifier "sum"
                                                       , quanVar = Identifier "__rules_repr_Set_Gent_repr_i"
                                                       , quanOver = Identifier "__rules_repr_Set_Gent_repr_t"
                                                       , quanGuard = Just (
                                                                       GenericNode
                                                                         Gt
                                                                         [ GenericNode
                                                                             Index
                                                                             [ Identifier "refn"
                                                                             , Identifier "__rules_repr_Set_Gent_repr_i"
                                                                             ]
                                                                         , ValueInteger 0
                                                                         ] )
                                                       , quanBody = ValueInteger 1
                                                       }
                                                   , Identifier "__rules_repr_Set_Gent_repr_mn"
                                                   ]
                                               , GenericNode
                                                   Leq
                                                   [ ExprQuantifier
                                                       { quanName = Identifier "sum"
                                                       , quanVar = Identifier "__rules_repr_Set_Gent_repr_i"
                                                       , quanOver = Identifier "__rules_repr_Set_Gent_repr_t"
                                                       , quanGuard = Just (
                                                                       GenericNode
                                                                         Gt
                                                                         [ GenericNode
                                                                             Index
                                                                             [ Identifier "refn"
                                                                             , Identifier "__rules_repr_Set_Gent_repr_i"
                                                                             ]
                                                                         , ValueInteger 0
                                                                         ] )
                                                       , quanBody = ValueInteger 1
                                                       }
                                                   , Identifier "__rules_repr_Set_Gent_repr_mx"
                                                   ]
                                               ] )
                    , reprCaseWheres = []
                    , reprCaseBindings = []
                    }
                , RuleReprCase
                    { reprCasePattern = DomainSet
                                          { size = Just ( Identifier "__rules_repr_Set_Gent_repr_n" )
                                          , minSize = Just ( Identifier "__rules_repr_Set_Gent_repr_mn" )
                                          , maxSize = Just ( Identifier "__rules_repr_Set_Gent_repr_mx" )
                                          , attrDontCare = False
                                          , element = Identifier "__rules_repr_Set_Gent_repr_t"
                                          , representation = Nothing
                                          }
                    , reprCaseStructural = Just (
                                             GenericNode
                                               Eq
                                               [ ExprQuantifier
                                                   { quanName = Identifier "sum"
                                                   , quanVar = Identifier "__rules_repr_Set_Gent_repr_i"
                                                   , quanOver = Identifier "__rules_repr_Set_Gent_repr_t"
                                                   , quanGuard = Just (
                                                                   GenericNode
                                                                     Gt
                                                                     [ GenericNode
                                                                         Index
                                                                         [ Identifier "refn"
                                                                         , Identifier "__rules_repr_Set_Gent_repr_i"
                                                                         ]
                                                                     , ValueInteger 0
                                                                     ] )
                                                   , quanBody = ValueInteger 1
                                                   }
                                               , Identifier "__rules_repr_Set_Gent_repr_n"
                                               ] )
                    , reprCaseWheres = []
                    , reprCaseBindings = []
                    }
                ]
  }