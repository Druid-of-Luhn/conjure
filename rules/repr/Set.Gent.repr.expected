
~~> Gent
~~> matrix indexed by [t] of int(0..mx)
~~> forall i : t . refn[i] = 0 \/ refn[i] = 1 + (sum j : t , j < i /\ refn[j] > 0 . 1)
    where t :: int

*** set of t
    letting mx be image(domSize, t)

*** set (size n) of t
    ~~> (sum i : t , refn[i] > 0 . 1) = n
    letting mx be image(domSize, t)

*** set (minSize mn) of t
    ~~> (sum i : t , refn[i] > 0 . 1) >= mn
    letting mx be image(domSize, t)

*** set (maxSize mx) of t
    ~~> (sum i : t , refn[i] > 0 . 1) <= mx

*** set (size n, minSize mn) of t
    ~~> (sum i : t , refn[i] > 0 . 1) = n
    letting mx be image(domSize, t)

*** set (size n, maxSize mx) of t
    ~~> (sum i : t , refn[i] > 0 . 1) = n

*** set (minSize mn, maxSize mx) of t
    ~~> (sum i : t , refn[i] > 0 . 1) >= mn /\ (sum i : t , refn[i] > 0 . 1) <= mx

*** set (size n, minSize mn, maxSize mx) of t
    ~~> (sum i : t , refn[i] > 0 . 1) = n
